keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,326 @@
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The `Burstable` QoS level is used for any Pod that defines resource limits for at least one container, and doesn't define the same value for both the `request` and `limit` sections. In these Pods, a container which doesn't define any limit for a resource will effectively not be constrained in any way for that resource. Defining a limit for all of the containers in the pod has the implication that it is not possible to define a pod where the resources are controlled on the pod level. 
+
+This KEP proposes a way to define constraints on the Pod level, to allow for a smoother resource allocation strategy for `Burstable` pods.
+
+
+## Motivation
+
+
+Some workloads are deployed as pods which are comprised of multiple sidecar containers which are strongly coupled in terms of their task. Such containers communicate either across a shared filesystem, or the localhost network, and orchestrate some common task. 
+
+For example, consider the following Pod with the following structure:
+
+<pre>
+  pod
+   |
+   +-- container1 (main task)
+   |
+   +-- container2 (second level task)
+   |
+   +-- container3 (log handler)
+   |
+   +-- container4 (mesh sidecar)
+</pre>
+
+In some cases deploying a single container with all the tasks is not optimal and not always possible. Kubernetes is not aware of these tasks, and doesn't monitor them for failure, and is not able to manage the resources (cgroups) for each of them. By separating the different tasks to their own containers, the application is able to leverage Kubernetes to monitor the tasks. The current `Burstable` QoS implementation requires the pod to either not limit each individual container at all, or micro-manage the resources allocated to each and every container.
+
+This proposal suggests a middle ground, and suggests a way to make it possible to describe to Kubernetes how to limit the resource consumption of multiple containers in the pod at once, on the Pod level, instead of trying to micro-manage the resource limits on the container level. For workloads where the work performed is burstable, this proposal would make it easier to allow the low-level mechanisms available in the underlying operating-system to manage the resources required for the task.
+
+### Goals
+
+This proposal aims to:
+
+*  Allow a `Burstable` pod to define that memory and CPU resource limits should be set on the level of the Pod.
```

Comment:
The same arguments for the Burstable tier would apply to the Guaranteed tier.  I could see users that do not want to specify explicit requests/limits for each container, but just want to bound the pod with fixed resources.

Reply:
The current behavior for the Guaranteed tier is already exactly this. A pod is in the guaranteed QoS if and only if it always has a request equal to the limit for all the containers in the pod. If this is the case, the current code already sets the pod level cgroup to the sum of the limits for each of the resources.

But because in the Guaranteed case all of the containers also have a limit set to them, setting the pod level cgroup to any limit is meaningless. The pod level cgroup limit is by definition greater than any individual container, which means that the individual container limit would apply before the pod level limit would apply.